### PR TITLE
Fix inaccurate loss calculation

### PR DESCRIPTION
**Pull request by felixlaumon**
For #247.

This PR changes the way the losses are aggregated. Now it will keep track of the sample per batch and use that as the weights when calculating the average.

The current calculation is slightly inaccurate under these circumstances.

1. `batch_size` is very large or the number of batches is small
2. losses is very noisy
3. the last batch has very few samples

I do not see any relevant tests related to loss calculation so I haven't add any. If you can give me some pointers about where to add them, I can write some.

### CODE REVIEW

File: nolearn/lasagne/base.py [hunk 1]
```diff
@@ -582,30 +582,34 @@ def train_loop(self, X, y, epochs=None):
 
             t0 = time()
 
+            batch_train_sizes = []
             for Xb, yb in self.batch_iterator_train(X_train, y_train):
                 batch_train_loss = self.apply_batch_func(
                     self.train_iter_, Xb, yb)
-                train_losses.append(batch_train_loss)
+                train_losses.append(batch_train_loss[0])
```
Comment (cancan101) [inline nolearn/lasagne/base.py]: why are you indexing here?

Comment (felixlaumon) [inline nolearn/lasagne/base.py]: Good catch – I forgot to mention this in the description.

It was because `batch_train_loss` is actually a list. See https://github.com/dnouri/nolearn/blob/master/nolearn/lasagne/base.py#L502. I am not sure why `loss_train` is wrapped in a list there.

**Comment by dnouri** [comment]
Thanks!

File: nolearn/lasagne/base.py [hunk 2]
```diff
@@ -582,30 +582,34 @@ def train_loop(self, X, y, epochs=None):
...
+                batch_valid_sizes.append(len(Xb))
 
                 if self.custom_scores:
                     y_prob = self.apply_batch_func(self.predict_iter_, Xb)
                     for custom_scorer, custom_score in zip(self.custom_scores, custom_scores):
                         custom_score.append(custom_scorer[1](yb, y_prob))
 
-            avg_train_loss = np.mean(train_losses)
-            avg_valid_loss = np.mean(valid_losses)
+            avg_train_loss = np.average(train_losses, weights=batch_train_sizes)
```
Comment (felixlaumon) [inline nolearn/lasagne/base.py]: Got a problem with this line just now. `np.average` raises exception if `arr` and `weights` are empty (length of 0). That happens if you `ctrl-c` before the first batch completes.

I can check the length of the losses and return early if empty, but I am not sure if that's a good solution. Any thoughts?

Comment (dnouri) [inline nolearn/lasagne/base.py]: Looks like before we got a warning and a `nan` in that case.  I'm fine if you wanna add a condition, maybe `avg_train_loss = np.average(train_losses, weights=batch_train_sizes) if train_losses else np.nan`?

Comment (cancan101) [inline nolearn/lasagne/base.py]: Probably best to wrap predicate with len since truthiness of array is
undefined...
